xonsh/events.py

```diff
@@ -0,0 +1,101 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.doc("Comes with eggs")
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guaranteed.
+    """
+
+    @classmethod
+    def doc(cls, text):
+        cls.__doc__ = text
+
+    def __call__(self, func):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+
+        A decorator method is added to the handler, validator(). If a validator
+        function is added, it can filter if the handler will be considered. The
+        validator takes the same arguments as the handler. If it returns False,
+        the handler will not called or considered, as if it was not registered
+        at all.
+        """
+        #  Using Pythons "private" munging to minimize hypothetical collisions
+        func.__validator = None
+        self.add(func)
+
+        def validator(vfunc):
+            """
+            Adds a validator function to a handler to limit when it is considered.
+            """
+            func.__validator = vfunc
+        func.validator = validator
+
+        return func
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+        """
+        for handler in self:
+            if handler.__validator is not None and not handler.__validator(*pargs, **kwargs):
+                continue
+            yield handler(*pargs, **kwargs)
+
+    def fire(self, *pargs, **kwargs):
+        """
+        The simplest use case: Calls each handler in turn with the provided
+        arguments and ignore the return values.
+        """
+        for _ in self.calleach(*pargs, **kwargs):
+            pass
+
+    def until_true(self, *pargs, **kwargs):
+        """
+        Calls each handler until one returns something truthy.
+
+        Returns that truthy value.
+        """
+        for rv in self.calleach(*pargs, **kwargs):
+            if rv:
+                return rv
+
+    def until_false(self, *pargs, **kwargs):
+        """
+        Calls each handler until one returns something falsey.
+        """
+        for rv in self.calleach(*pargs, **kwargs):
+            if not rv:
+                return rv
+
+
+class Events:
```

Comment:
I think Events is correct here, since it does hold and manage many Event instances.  Also, I'd rather we not get too caught up on bike shedding the names while we still have core architecture to work out :).